instrumentation/gopkg.in/olivere/elastic/splunkelastic/elastic.go

```diff
@@ -0,0 +1,157 @@
+func WrapRoundTripper(rt http.RoundTripper, opts ...Option) http.RoundTripper {
+	if rt == nil {
+		rt = http.DefaultTransport
+	}
+
+	cfg := internal.NewConfig(instrumentationName, localToInternal(opts)...)
+	cfg.DefaultStartOpts = append([]trace.SpanStartOption{
+		trace.WithSpanKind(trace.SpanKindClient),
+		trace.WithAttributes(semconv.DBSystemElasticsearch),
+	}, cfg.DefaultStartOpts...)
+
+	return &roundTripper{RoundTripper: rt, cfg: cfg}
+}
+
+// roundTripper wraps an http.RoundTripper's requests with a span.
+type roundTripper struct {
+	http.RoundTripper
+
+	cfg *internal.Config
+}
+
+var _ http.RoundTripper = (*roundTripper)(nil)
+
+func (rt *roundTripper) RoundTrip(r *http.Request) (resp *http.Response, err error) {
+	opts := rt.cfg.MergedSpanStartOptions(
+		trace.WithAttributes(semconv.HTTPClientAttributesFromHTTPRequest(r)...),
+		trace.WithAttributes(semconv.NetAttributesFromHTTPRequest("tcp", r)...),
+	)
+
+	tracer := rt.cfg.ResolveTracer(r.Context())
+	ctx, span := tracer.Start(r.Context(), name(r), opts...)
+
+	// Ensure anything downstream knows about the started span.
+	r = r.WithContext(ctx)
+	rt.cfg.Propagator.Inject(ctx, propagation.HeaderCarrier(r.Header))
+
+	resp, err = rt.RoundTripper.RoundTrip(r)
+	if err != nil {
+		span.RecordError(err)
+		span.SetStatus(codes.Error, err.Error())
+		span.End()
+		return
+	}
+
+	span.SetAttributes(semconv.HTTPAttributesFromHTTPStatusCode(resp.StatusCode)...)
+	span.SetStatus(semconv.SpanStatusFromHTTPStatusCode(resp.StatusCode))
+	resp.Body = &wrappedBody{ctx: ctx, span: span, body: resp.Body}
+
+	return
+}
+
+// name returns an appropriate span name based on the client request.
+// OpenTelemetry semantic conventions require this name to be low cardinality,
+// but since the Elasticsearch API is somewhat predictable we can usually
+// return more than just "HTTP {METHOD}". If this is a recognized
+// Elasticsearch operation the returned span name will conform with
+// OpenTelemetry database semantics, otherwise HTTP semantics will be used.
+func name(r *http.Request) string {
+	path := r.URL.Path
+	if path == "" {
+		path = "/"
+	}
+
+	tokenized := tokenize(path)
+	if tokenized == "" {
+		// Unrecognized Elasticsearch path, default to HTTP semantics.
+		return "HTTP " + r.Method
+	}
+
+	op, ok := operations[url{method: r.Method, path: tokenized}]
+	if !ok {
+		// Unrecognized Elasticsearch operation, default to HTTP semantics.
+		return "HTTP " + r.Method + " " + tokenized
+	}
+
+	if strings.HasPrefix(tokenized, "/{index}") {
+		// Use the {index} as the OpenTelemetry semantic for DB name.
+		p := strings.TrimPrefix(path, "/")
+		// Either: ["example-index"] or ["example-index", "*"]
+		const nParts = 2
+		idx := strings.SplitN(p, "/", nParts)[0]
+		// <db.operation> <db.name>
+		return op + " " + idx
+	}
+
+	// <db.operation>
+	return op
+}
+
+type wrappedBody struct {
+	ctx  context.Context
+	span trace.Span
+	body io.ReadCloser
+}
+
```

Comment:
Interesting technique. Does this mean the span will never end if the full body is not read or reader is not closed? I imagine any correctly written code would always call Close but it is rare for an instrumentation to rely on the end user writing correct code to function properly.

Reply:
The same problem is with `http` package instrumentation. Probably we could use [`runtime.SetFinalizer`](https://pkg.go.dev/runtime#SetFinalizer) to call Close, but the overhead may be big... Maybe it could be an opt-out functionality?

Reply:
More: https://medium.com/a-journey-with-go/go-finalizers-786df8e17687 
I just learned that they are used in `http` package anyway so probably this is how we should handle the scenario if the instrument code does not close the Body

Reply:
However, it won't help a lot as the span times would be VERY inaccurate...

Reply:
> Interesting technique. Does this mean the span will never end if the full body is not read or reader is not closed? I imagine any correctly written code would always call Close but it is rare for an instrumentation to rely on the end user writing correct code to function properly.

Yeah, as pointed out I had copied this from the upstream HTTP instrumentation (and from our local HTTP instrumentation). But I think this comes down to how we want to model the telemetry. For HTTP libraries including the response parsing in the span timing seems appropriate, especially if you want body parsing errors to fail the span. With Elasticsearch database telemetry this might not apply. Modeling the span around the connection and remote procedure itself seems to fit better with our other database instrumentation.

I'll look into updating this to match the database model. Sound good?

Reply:
Sounds good!